src/validators/union.rs

```diff
@@ -106,3 +111,127 @@ impl Validator for UnionValidator {
         self.choices.iter_mut().try_for_each(|v| v.complete(build_context))
     }
 }
+
+#[derive(Debug, Clone)]
+pub struct TaggedUnionValidator {
+    choices: AHashMap<String, CombinedValidator>,
+    lookup_key: LookupKey,
+    from_attributes: bool,
+    strict: bool,
+    tags_repr: String,
+    key_repr: String,
+    name: String,
+}
+
+impl BuildValidator for TaggedUnionValidator {
+    const EXPECTED_TYPE: &'static str = "tagged-union";
+
+    fn build(
+        schema: &PyDict,
+        config: Option<&PyDict>,
+        build_context: &mut BuildContext,
+    ) -> PyResult<CombinedValidator> {
+        let py = schema.py();
+        let lookup_key = match LookupKey::from_py(py, schema, None, "tag_key", "tag_keys")? {
+            Some(lookup_key) => lookup_key,
+            None => return py_error!(PyKeyError; "'tag_key' or 'tag_keys' must be set on a tagged union"),
+        };
+
+        let mut choices = AHashMap::new();
+        let mut first = true;
+        let mut tags_repr = String::with_capacity(50);
+        let mut descr = String::with_capacity(50);
+
+        for item in schema.get_as_req::<&PyDict>("choices")?.items().iter() {
+            let tag: String = item.get_item(0)?.extract()?;
+            let value = item.get_item(1)?;
+            let validator = build_validator(value, config, build_context)?.0;
+            if first {
+                first = false;
+                write!(tags_repr, r#""{}""#, tag).unwrap();
+                descr.push_str(validator.get_name());
+            } else {
+                write!(tags_repr, r#", "{}""#, tag).unwrap();
+                // no spaces in get_name() output to make loc easy to read
+                write!(descr, ",{}", validator.get_name()).unwrap();
+            }
+            choices.insert(tag, validator);
+        }
+
+        let from_attributes =
+            schema_or_config(schema, config, "from_attributes", "typed_dict_from_attributes")?.unwrap_or(false);
+        let key_repr = lookup_key.to_string();
```

Comment:
Is it needed?

Reply:
[bc63a63](https://github.com/samuelcolvin/pydantic-core/pull/157/commits/bc63a63350074530820a09cf2a82ff02d950d20b)

Reply:
Humm, have you benchmarked?

Generally if we can prepare stuff like this during build, I would prefer to.

Even if the individual effect is small, they might all add up to a few percent.

Reply:
I thought avoid an extra string allocation (less memory) for each tagged-union validator in all cases was worth the operation only in case of errors.

Reply:
You might be right, I haven't benchmarked.

Reply:
If the slowdown is <1% with all errors let's go for it.